File: bindings/imgui_bundle/demos_python/demos_immapp/demo_drag_and_drop.py
# Note: the drag and drop API differs a bit between C++ and Python.
# * In C++, ImGui::SetDragDropPayload and AcceptDragDropPayload are able to accept any kind of object
#   (by storing a buffer whose size is the object size).
#
# Unfortunately, this behaviour cannot be reproduced in python.
#
# * In Python, you can use imgui.set_drag_drop_payload_py_id and imgui.accept_drag_drop_payload_py_id.
#   These versions can only store an integer id for the payload
#   (so that you may have to store the corresponding payload somewhere else)
import enum
from typing import List
from imgui_bundle import imgui, immapp


class DragMode(enum.Enum):
    Copy = enum.auto()
    Move = enum.auto()
    Swap = enum.auto()


class DemoState:
    mode: DragMode
    names: List[str]

    def __init__(self):
        self.mode = DragMode.Copy
        self.names = [
            "Bobby",
            "Beatrice",
            "Betty",
            "Brianna",
            "Barry",
            "Bernard",
            "Bibi",
            "Blaine",
            "Bryn",
        ]


def gui_drag_and_drop(state: DemoState) -> None:
    if imgui.radio_button("Copy", state.mode == DragMode.Copy):
        state.mode = DragMode.Copy
    imgui.same_line()
    if imgui.radio_button("Move", state.mode == DragMode.Move):
        state.mode = DragMode.Move
    imgui.same_line()
    if imgui.radio_button("Swap", state.mode == DragMode.Swap):
        state.mode = DragMode.Swap

    for n in range(len(state.names)):
        imgui.push_id(n)
        if n % 3 != 0:
            imgui.same_line()
        imgui.button(state.names[n], immapp.em_to_vec2(5.0, 5.0))

        # Our buttons are both drag sources and drag targets here!
        if imgui.begin_drag_drop_source(imgui.DragDropFlags_.none.value):
            # Set payload to carry the index of our item (in python, the payload is an int)
            drag_data_id = n
            imgui.set_drag_drop_payload_py_id("DND_DEMO_CELL", drag_data_id)

            # Display preview (could be anything, e.g. when dragging an image we could decide to display
            # the filename and a small preview of the image, etc.)
            if state.mode == DragMode.Copy:
                imgui.text(f"Copy {state.names[n]}")
            elif state.mode == DragMode.Move:
                imgui.text(f"Move {state.names[n]}")
            elif state.mode == DragMode.Swap:
                imgui.text(f"Swap {state.names[n]}")
            imgui.end_drag_drop_source()

        if imgui.begin_drag_drop_target():
            payload = imgui.accept_drag_drop_payload_py_id("DND_DEMO_CELL")
            if payload is not None:
                payload_n = payload.data_id
                if state.mode == DragMode.Copy:
                    state.names[n] = state.names[payload_n]
                elif state.mode == DragMode.Move:
                    state.names[n] = state.names[payload_n]
                    state.names[payload_n] = ""
                elif state.mode == DragMode.Swap:
                    state.names[n], state.names[payload_n] = (
                        state.names[payload_n],
                        state.names[n],
                    )
            imgui.end_drag_drop_target()
        imgui.pop_id()


def main() -> None:
    state = DemoState()

    def gui():
        gui_drag_and_drop(state)

    immapp.run(gui)


if __name__ == "__main__":
    main()

File: bindings/imgui_bundle/demos_cpp/demos_immapp/demo_drag_and_drop.cpp
#include "imgui.h"
#include "immapp/immapp.h"
#include <vector>
#include <string>


enum class DragMode
{
    Copy,
    Move,
    Swap
};


struct DemoState
{
    DragMode Mode = DragMode::Copy;
    std::vector<std::string> Names = {
        "Bobby", "Beatrice", "Betty",
        "Brianna", "Barry", "Bernard",
        "Bibi", "Blaine", "Bryn"
    };
};


void gui_drag_and_drop(DemoState& state)
{
    if (ImGui::RadioButton("Copy", state.Mode == DragMode::Copy))
        state.Mode = DragMode::Copy;
    ImGui::SameLine();
    if (ImGui::RadioButton("Move", state.Mode == DragMode::Move))
        state.Mode = DragMode::Move;
    ImGui::SameLine();
    if (ImGui::RadioButton("Swap", state.Mode == DragMode::Swap))
        state.Mode = DragMode::Swap;

    for (int n = 0; n < state.Names.size(); n++)
    {
        ImGui::PushID(n);
        if ((n % 3) != 0)
            ImGui::SameLine();
        ImGui::Button(state.Names[n].c_str(), ImmApp::EmToVec2(5.f, 5.f));

        // Our buttons are both drag sources and drag targets here!
        if (ImGui::BeginDragDropSource(ImGuiDragDropFlags_None))
        {
            // Set payload to carry the index of our item (could be anything)
            ImGui::SetDragDropPayload("DND_DEMO_CELL", &n, sizeof(int));

            // Display preview (could be anything, e.g. when dragging an image we could decide to display
            // the filename and a small preview of the image, etc.)
            if (state.Mode == DragMode::Copy) { ImGui::Text("Copy %s", state.Names[n].c_str()); }
            if (state.Mode == DragMode::Move) { ImGui::Text("Move %s", state.Names[n].c_str()); }
            if (state.Mode == DragMode::Swap) { ImGui::Text("Swap %s", state.Names[n].c_str()); }
            ImGui::EndDragDropSource();
        }
        if (ImGui::BeginDragDropTarget())
        {
            if (const ImGuiPayload* payload = ImGui::AcceptDragDropPayload("DND_DEMO_CELL"))
            {
                IM_ASSERT(payload->DataSize == sizeof(int));
                int payload_n = *(const int*)payload->Data;
                if (state.Mode == DragMode::Copy)
                {
                    state.Names[n] = state.Names[payload_n];
                }
                if (state.Mode == DragMode::Move)
                {
                    state.Names[n] = state.Names[payload_n];
                    state.Names[payload_n] = "";
                }
                if (state.Mode == DragMode::Swap)
                {
                    std::string tmp = state.Names[n];
                    state.Names[n] = state.Names[payload_n];
                    state.Names[payload_n] = tmp;
                }
            }
            ImGui::EndDragDropTarget();
        }
        ImGui::PopID();
    }

}


int main(int, char**)
{
    DemoState state;
    auto gui = [&state]() { gui_drag_and_drop(state); };
    ImmApp::Run(gui);
}
